Plugin typings

streamich opened the pull request:
## Summary

Adds TypeScript typings to legacy plugin functions. So they can be annotated like this:

```
const apmOss: LegacyPlugin = kibana => {
  // ...
};
```

Autocompletion:

![image](https://user-images.githubusercontent.com/9773803/57524454-1d5c8700-7328-11e9-8a99-afa3ca345d24.png)


In this PR I did not want to type annotate the whole legacy plugin system, but at least the most necessary part to get the ball rolling. (Added type annotation to APM plugin just because it is the first plugin in `x-pack`.)

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- ~~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
- ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- ~~[Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~~
- ~~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

streamich committed:
feat: 🎸 add typings for legacy plugin factory function

streamich committed:
feat: 🎸 annotate APM plugin with LegacyPlugin and fix ret error

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13531/)

spalger commented on src/legacy/plugin_discovery/types.ts:
```diff
@@ -0,0 +1,87 @@
+  isEnabled: boolean;
+}
+
+export interface UiExports {
```
This is unused, and incomplete, so I think it's fine to discard it.

streamich replied:
I'm using it now and improved it.

spalger reviewed (approved):
One little nit, but LGTM

joshdover commented on src/legacy/plugin_discovery/types.ts:
````diff
@@ -0,0 +1,87 @@
+ * };
+ * ```
+ */
+export type LegacyPlugin = (kibana: LegacyPluginApi) => ArrayOrItem<LegacyPluginSpec>;
````
optional: I think a better name for this might be `LegacyPluginInitializer` which would be analogous to the new platform's `PluginInitializer` type.

streamich replied:
Done.

streamich committed:
feat: 🎸 address review comments

Rename LegacyPlugin to LegacyPluginInitializer and improve uiExports
annotations.

spalger reviewed (approved):
LGTM

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13569/)
